### PR TITLE
[C++ API] InstanceNorm{1,2,3}d

### PR DESCRIPTION
**Pull request by divyanshsinghvi**
Hi @yf225, 

I have a few doubts related to implementation:
1) What tests do I have to write? 
2) What does _load_state_from_dict does? 
3) Do I need to override reset() function as I can not see it's utility?
4) InstanceNormOptions could be removed with BatchNormOptions, but I find that 
`track_running_status` is not defined instead `stateful` is defined. 

InstanceNorm{1,2,3}d #25883

### CODE REVIEW

**Comment by SsnL** [comment]
`_load_state_from_dict` is used to resolve BC issue for loading old python instance norm state dict, because the state dict of instance norm changed at one point. I believe that it is not relevant for CPP API.

**Review by yf225 — COMMENTED** [review]
@divyanshsinghvi Thanks so much for the awesome work! I believe we will need to rebase this on top of https://github.com/pytorch/pytorch/pull/28176 after it's merged, because https://github.com/pytorch/pytorch/pull/28176 fixes the `track_running_status` vs. `stateful` issue that you mentioned, and also creates `BatchNormImplBase` which we'll need to subclass `InstanceNormImpl` from.

**Comment by yf225** [comment]
For this PR we shouldn't need to override `reset` - we will get more clarity about it once we are able to rebase on top of https://github.com/pytorch/pytorch/pull/28176 :D

**Comment by yf225** [comment]
@divyanshsinghvi I just rebased this PR on top of upstream master which contains the changes in https://github.com/pytorch/pytorch/pull/28176. I think we can let `InstanceNormImpl` subclass from the new `BatchNormImplBase` class, which mirrors the class hierarchy in Python. I also think we do need to define a separate `InstanceNormOptions` struct, because `InstanceNorm` constructor's `affine` and `track_running_stats` parameters take a different default value than the `BatchNorm` constructor's `affine` and `track_running_stats` parameters.

**Comment by yf225** [comment]
I think it should be sufficient to provide an empty override for the `reset()` function, because I believe not overriding it would cause some build errors.

**Comment by yf225** [comment]
For tests, I think we don't need to be comprehensive and can just check for common input / output. This is an example for `F::batch_norm` https://github.com/pytorch/pytorch/blob/a8b63cacbca5ef83bc105488c6b5cbc7cfa10537/test/cpp/api/functional.cpp#L1269-L1294 and we can do the same for the module tests as well https://github.com/pytorch/pytorch/blob/a8b63cacbca5ef83bc105488c6b5cbc7cfa10537/test/cpp/api/modules.cpp#L1136-L1147

**Comment by divyanshsinghvi** [comment]
> I think it should be sufficient to provide an empty override for the `reset()` function, because I believe not overriding it would cause `InstanceNormImpl` to become an abstract class, and we cannot create objects of abstract classes.

Sorry I might be wrong in my understanding here. But since we are using `BatchNormImplBase` to subclass from, it would have already overridden the reset function. Hence, `InstanceNormImpl` shouldn't be an abstract class due to this. Though, it is likely the case that we would not like the function of `reset` defined in `BatchNormImplBase` getting called and hence we should redefine it

**Comment by yf225** [comment]
@divyanshsinghvi Thanks a lot for the investigation! Yes I think you are right and we don't need to override `reset` in `InstanceNormImpl`, because it's already overridden in `BatchNormImplBase`.

File: torch/csrc/api/include/torch/nn/functional/instancenorm.h [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#pragma once
+
+namespace torch {
+namespace nn {
+namespace functional {
+
+inline Tensor instance_norm(const Tensor& input, const Tensor& running_mean,
+    const Tensor& running_var, const Tensor& weight, const Tensor& bias, 
+    bool use_input_stats, double momentum, double eps) {
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/instancenorm.h]: I think we can put the parameters
```python
running_mean=None, running_var=None, weight=None,
bias=None, use_input_stats=True, momentum=0.1, eps=1e-5
```
into `InstanceNormOptions` as well, so that the functional form can be:
```cpp
inline Tensor instance_norm(const Tensor& input, const InstanceNormOptions& options = {}) {
```

File: torch/csrc/api/include/torch/nn/options/instancenorm.h [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+#include <torch/arg.h>
+#include <torch/csrc/WindowsTorchApiMacro.h>
+#include <torch/types.h>
+
+namespace torch {
+namespace nn {
+
+struct TORCH_API InstanceNormOptions {
+  /* implicit */ InstanceNormOptions(int64_t num_features);
+  TORCH_ARG(int64_t, num_features);
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/options/instancenorm.h]: Since `InstanceNormOptions` is used in the functional form and it doesn't need the `num_features` parameter, we can make `num_features` an object of `c10::optional<int64_t>` and make its default value `c10::nullopt`, and add a default constructor for `InstanceNormOptions` so that the functional form's signature can be
```cpp
inline Tensor instance_norm(const Tensor& input, const InstanceNormOptions& options = {}) {
```

**Review by yf225 — COMMENTED** [review]
@divyanshsinghvi Thanks so much for the awesome work! This work is highly complex and thanks so much for your help on this. I left some comments that we can work through together.

File: torch/csrc/api/include/torch/nn/modules/instancenorm.h [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  Tensor forward(const Tensor& input);
+  
+  /// The options with which this `Module` was constructed.
+  InstanceNormOptions options;
+};
+
+class TORCH_API InstanceNorm1dImpl : public InstanceNormImpl<1, InstanceNorm1dImpl, BatchNorm1dImpl> {
+ public:
+  using InstanceNormImpl<1, InstanceNorm1dImpl, BatchNorm1dImpl>::InstanceNormImpl;
+ private:
```
[the remark this answers is not in the document]
Comment (divyanshsinghvi) [reply torch/csrc/api/include/torch/nn/modules/instancenorm.h]: Yes, right.

**Comment by divyanshsinghvi** [comment]
Hi @yf225, 
Is there something I could do after your current bug-fix commits. I was on a long vacation and could commit only once. Thanks for handling meanwhile.

**Comment by yf225** [comment]
@divyanshsinghvi No worries about it - we have an internal branch cut deadline next week, and I will make the changes and add tests in order to get this PR in before the deadline. Thanks a lot for your help!

**Review by yf225 — APPROVED** [review]
@divyanshsinghvi Thanks a lot for the awesome work!
